### ISSUE TITLE
Specify -Wdeprecated-declarations for proto targets

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#14281]
The deprecated proto fields are generating a whole lot of warnings:

```
bazel-out/k8-fastbuild/bin/external/envoy_api/envoy/config/core/v3/config_source.pb.h:102:3: note: declared here
  102 |   ApiConfigSource_ApiType_hidden_envoy_deprecated_UNSUPPORTED_REST_LEGACY PROTOBUF_DEPRECATED_ENUM = 0,
```

We should set the `-Wdeprecated-declarations` for proto targets to suppress this warning.

**Comment by mattklein123** [comment]
I noticed this also. I took a very quick look and I thought we already do disable it when compiling cc libraries:
https://github.com/envoyproxy/envoy/blob/e7323c3a5ff9f00d2a97af7788bb654952ced7d1/bazel/envoy_internal.bzl#L17

But maybe this is not used when compiling proto generated code? I find that strange though.

cc @lizan @htuch?

**Comment by snowp** [comment]
I don't think we pass `envoy_copts` to the API targets, just the regular `envoy_cc_library` targets. I'm guessing there's just some target that needs to be updated to include `envoy_copts` (or a different set of flags, not sure).

**Comment by mattklein123** [comment]
It seems like a bug that we don't use envoy_copts to compile the proto targets?

**Comment by htuch** [comment]
The copts used by the APIs come from `api/bazel`, which, for reasons of `data-plane-api` mirroring, was part of the standalone `api/` tree. We should probably push these options there. It's possible also, with some Bazel magic, to append envoy_copts during WORKSPACE inclusion.

**Comment by mattklein123** [comment]
> It's possible also, with some Bazel magic, to append envoy_copts to the API opts (I think..) during WORKSPACE inclusion.

This might be a better option but I will leave it to you build gurus to sort out. :)

**Comment by htuch** [comment]
Actually, I think PGV creates this rule, https://github.com/envoyproxy/envoy/blob/master/api/bazel/api_build_system.bzl#L125.

**Comment by mattklein123** [comment]
Yeah I saw that. Would it be best to just plumb through being able to pass in copts into those rules and then we can pass the main Envoy ones in? We really should be compiling the protos with the same compile options as the rest of the code.

**Comment by htuch** [comment]
I think the easiest way to do this might be to factor out compile options to its own .bzl, and then copy it in with the repository rule in https://github.com/envoyproxy/envoy/blob/61d80ba86c1d9e01072a1d9b14e8b84c4229b919/bazel/api_binding.bzl#L15.
